{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Acquisition of the NEGATIVE dataset via PhANNs DB\n",
    "PhANNs DB article link: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC7660903/\n",
    "\n",
    "To determine whether anti-CRISPR proteins would be encoded as structural proteins and thus included in the \"classes\" dataset of PhANNs DB, we need to understand the nature of anti-CRISPR proteins and how they function within bacteriophages.\n",
    "\n",
    "Anti-CRISPR proteins are typically produced by bacteriophages. These proteins inhibit the activity of the CRISPR-Cas system. Anti-CRISPR proteins are not typically considered structural components of the phage particle itself. Instead, they are regulatory proteins that interfere with the host's defense mechanisms. Therefore, it's unlikely that anti-CRISPR proteins would be classified as structural proteins in the context of the \"classes\" dataset in PhANNs DB. Given this, the negative dataset will be obtained from the \"classes\" database of PhANNs DB. Since anti-CRISPR proteins are regulatory in nature and not structural, we will select the structural proteins as the negative dataset.\n",
    "\n",
    "The link to the file downloaded from the PhANNs DB website is http://phanns.com/download/expandedDB.tgz. This is the chosen file as it has already been preprocessed by the authors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "from Bio import SeqIO\n",
    "import pandas as pd\n",
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Combined FASTA file saved as ProducedDatasets/negative_dataset.fasta\n"
     ]
    }
   ],
   "source": [
    "# Conversion of selected .fasta files into a single .fasta file\n",
    "def merge_fasta(input_dir, output_dir, output_file):\n",
    "    output_path = os.path.join(output_dir, output_file)\n",
    "    with open(output_path, \"w\") as out_handle:\n",
    "        for filename in os.listdir(input_dir):\n",
    "            if filename.endswith(\".fasta\") and not filename.endswith(\"_other.fasta\"): # Selecting all .fasta files EXCEPT \"_other.fasta\" files\n",
    "                filepath = os.path.join(input_dir, filename)\n",
    "                for record in SeqIO.parse(filepath, \"fasta\"):\n",
    "                    SeqIO.write(record, out_handle, \"fasta\")    # Writes the combined fasta file\n",
    "\n",
    "# Directory containing input FASTA files (relative path)\n",
    "input_dir = \"PhANNs/expandedDB\"\n",
    "\n",
    "# Output directory (current working directory)\n",
    "output_dir = os.getcwd()\n",
    "\n",
    "# Output file\n",
    "output_file = \"ProducedDatasets/negative_dataset.fasta\"\n",
    "\n",
    "# Merge the input FASTA files into one\n",
    "merge_fasta(input_dir, output_dir, output_file)\n",
    "\n",
    "print(\"Combined FASTA file saved as\", os.path.join(output_file))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Reads the FASTA file created and converts it into a pandas dataframe (clustered_negative_cdhit.fasta)\n",
    "def fasta2df(fasta_file):\n",
    "    \"\"\"\n",
    "    Reads the CD-HIT .fasta file output of the dataset and converts it into a pandas DataFrame.\n",
    "\n",
    "    Parameters:\n",
    "    fasta_file (str): Path to the FASTA file of the negative dataset.\n",
    "\n",
    "    Returns:\n",
    "    pd.DataFrame: DataFrame with columns:\n",
    "        - 'ID': Accession IDs of the sequences.\n",
    "        - 'Sequence': Corresponding protein sequences.\n",
    "        - 'Organism': Organism information extracted from the sequence description.\n",
    "        - 'Structure': Description of the protein structure.\n",
    "        - 'Size': Length of the protein sequences.\n",
    "        - 'Protein Acr': A column with the value 'No' for all entries.\n",
    "    \"\"\"\n",
    "    # Read the FASTA file and extract required information for each record\n",
    "    data = [(fasta.id,\n",
    "             str(fasta.seq),\n",
    "             re.search(r'\\[(.*?)\\]', fasta.description).group(1),\n",
    "             re.sub(r'\\[.*?\\]', '', fasta.description).strip().replace(fasta.id, '').strip())\n",
    "            for fasta in SeqIO.parse(open(fasta_file), 'fasta')]\n",
    "\n",
    "    # Create a DataFrame from the extracted data\n",
    "    df = pd.DataFrame(data, columns=['ID', 'Sequence', 'Organism', 'Structure'])\n",
    "\n",
    "    # Add new columns 'Size' and 'Protein Acr'\n",
    "    df['Size'] = df['Sequence'].apply(len)\n",
    "    df['Protein Acr'] = 'No'\n",
    "\n",
    "    return df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>Sequence</th>\n",
       "      <th>Organism</th>\n",
       "      <th>Structure</th>\n",
       "      <th>Size</th>\n",
       "      <th>Protein Acr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>WP_132722810.1</td>\n",
       "      <td>MTNNTDTFLHYREGKTQSQRFLKNLDPENLKLNDLDVADWLLFAFN...</td>\n",
       "      <td>Tenacibaculum sp. M341</td>\n",
       "      <td>phage baseplate protein</td>\n",
       "      <td>1042</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>WP_130915135.1</td>\n",
       "      <td>MKKTDTFSHYREGKSQMQRFLAELDPGNLELHDFDLFDWLLFANNF...</td>\n",
       "      <td>Chryseobacterium gleum</td>\n",
       "      <td>phage baseplate protein</td>\n",
       "      <td>1027</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AYZ12950.1</td>\n",
       "      <td>MKKTDTFSHYREGKSQMQRFLAELDPGNLELHDFDLFDWLLFANNF...</td>\n",
       "      <td>Chryseobacterium arthrosphaerae</td>\n",
       "      <td>phage baseplate protein</td>\n",
       "      <td>1016</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>WP_124535473.1</td>\n",
       "      <td>MKKTDTFSHYREGKSQMQRFLAELDPSNLELHDFDLFDWLLFANNF...</td>\n",
       "      <td>Chryseobacterium sp. KBW03</td>\n",
       "      <td>phage baseplate protein</td>\n",
       "      <td>1017</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>WP_123887321.1</td>\n",
       "      <td>MKKTDTFSHYREGKSQMQRFLAELDPGNLELHDFDLFDWLLFANNF...</td>\n",
       "      <td>Chryseobacterium indoltheticum</td>\n",
       "      <td>phage baseplate protein</td>\n",
       "      <td>1016</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168655</th>\n",
       "      <td>YP_009031970.1</td>\n",
       "      <td>MAMNAHTPFDAKQDWTNPYCQNSSNDPMVDALLGNAYHVVRTVYCN...</td>\n",
       "      <td>Escherichia phage Bp4</td>\n",
       "      <td>tail fiber protein</td>\n",
       "      <td>600</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168656</th>\n",
       "      <td>AGC31500.1</td>\n",
       "      <td>MNAHTPFDAKKDWSNPYCQNSSNDPMVDALLGNAYHVVRTVYCNLG...</td>\n",
       "      <td>Escherichia phage EC1-UPM</td>\n",
       "      <td>tail fiber protein</td>\n",
       "      <td>928</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168657</th>\n",
       "      <td>AEL79676.1</td>\n",
       "      <td>MNAHTPFDANQDWSNPYCQNKSNDPMVDALLGNAYHVVRTVYCNLG...</td>\n",
       "      <td>Escherichia phage vB_EcoP_G7C</td>\n",
       "      <td>tail fiber protein</td>\n",
       "      <td>1063</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168658</th>\n",
       "      <td>EHJ79759.1</td>\n",
       "      <td>MDNEFYTLLTDRGMAKIASALADKKQIHLQKMAVGDGGGQYYEPTA...</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ba...</td>\n",
       "      <td>Phage tail fiber protein</td>\n",
       "      <td>476</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>168659</th>\n",
       "      <td>EGG93505.1</td>\n",
       "      <td>MANLPENAEFTEGIYQLEQLDPVHGGPDGIDNLQAKQLANRTVWLK...</td>\n",
       "      <td>gamma proteobacterium IMCC1989</td>\n",
       "      <td>Phage tail fiber protein</td>\n",
       "      <td>247</td>\n",
       "      <td>No</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>168660 rows Ã— 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    ID                                           Sequence  \\\n",
       "0       WP_132722810.1  MTNNTDTFLHYREGKTQSQRFLKNLDPENLKLNDLDVADWLLFAFN...   \n",
       "1       WP_130915135.1  MKKTDTFSHYREGKSQMQRFLAELDPGNLELHDFDLFDWLLFANNF...   \n",
       "2           AYZ12950.1  MKKTDTFSHYREGKSQMQRFLAELDPGNLELHDFDLFDWLLFANNF...   \n",
       "3       WP_124535473.1  MKKTDTFSHYREGKSQMQRFLAELDPSNLELHDFDLFDWLLFANNF...   \n",
       "4       WP_123887321.1  MKKTDTFSHYREGKSQMQRFLAELDPGNLELHDFDLFDWLLFANNF...   \n",
       "...                ...                                                ...   \n",
       "168655  YP_009031970.1  MAMNAHTPFDAKQDWTNPYCQNSSNDPMVDALLGNAYHVVRTVYCN...   \n",
       "168656      AGC31500.1  MNAHTPFDAKKDWSNPYCQNSSNDPMVDALLGNAYHVVRTVYCNLG...   \n",
       "168657      AEL79676.1  MNAHTPFDANQDWSNPYCQNKSNDPMVDALLGNAYHVVRTVYCNLG...   \n",
       "168658      EHJ79759.1  MDNEFYTLLTDRGMAKIASALADKKQIHLQKMAVGDGGGQYYEPTA...   \n",
       "168659      EGG93505.1  MANLPENAEFTEGIYQLEQLDPVHGGPDGIDNLQAKQLANRTVWLK...   \n",
       "\n",
       "                                                 Organism  \\\n",
       "0                                  Tenacibaculum sp. M341   \n",
       "1                                  Chryseobacterium gleum   \n",
       "2                         Chryseobacterium arthrosphaerae   \n",
       "3                              Chryseobacterium sp. KBW03   \n",
       "4                          Chryseobacterium indoltheticum   \n",
       "...                                                   ...   \n",
       "168655                              Escherichia phage Bp4   \n",
       "168656                          Escherichia phage EC1-UPM   \n",
       "168657                      Escherichia phage vB_EcoP_G7C   \n",
       "168658  Salmonella enterica subsp. enterica serovar Ba...   \n",
       "168659                     gamma proteobacterium IMCC1989   \n",
       "\n",
       "                       Structure  Size Protein Acr  \n",
       "0        phage baseplate protein  1042          No  \n",
       "1        phage baseplate protein  1027          No  \n",
       "2        phage baseplate protein  1016          No  \n",
       "3        phage baseplate protein  1017          No  \n",
       "4        phage baseplate protein  1016          No  \n",
       "...                          ...   ...         ...  \n",
       "168655        tail fiber protein   600          No  \n",
       "168656        tail fiber protein   928          No  \n",
       "168657        tail fiber protein  1063          No  \n",
       "168658  Phage tail fiber protein   476          No  \n",
       "168659  Phage tail fiber protein   247          No  \n",
       "\n",
       "[168660 rows x 6 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_negative = fasta2df('ProducedDatasets/negative_dataset.fasta')\n",
    "df_negative "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
