# Bioinformatics Project pertaining to the degree of **Master's in Bioinformatics**, University of Minho
## PhageAcr: Identification of anti-CRISPR proteins

Project developed by Christian Neitzel (PG52526) under the supervision of Fernanda Vieira, Hugo Oliveira and Óscar Dias from the Centre of Biological Engineering of the University of Minho.

This project was evaluated in three phases:
**1) Phase 1:** Delivery of an initial article (file "Projeto_PhageAcr_ChristianNeitzel_ArtigoIntercalar.pdf") containing the context of the project to be developed, the state-of-the-art, objectives and potential tools that could provide insights into how the project could be developed;
**2) Phase 2:** Presentation via PowerPoint of the project detailing the chosen methodologies, results obtained and conclusions made;
**3) Phase 3:** Delivery of a final article containing all aforementioned contextualizations, methods, results, conclusions and future perspectives.


### Project description:

Bacteriophages (phages) are the most abundant life form in the planet. They are bacterial viruses that recognize, infect and kill their hosts to propagate. As phages are very different from each other and from bacteria, and we have relatively few phage genes in our database compared to bacterial genes, we are unable to assign function to 50–90% of phage genes.

**Anti-CRISPRs** (Acr) are short proteins (<200 aa) encoded by some phages to overcome bacterial host immunity. In particular, they inhibit the CRISPR-Cas (clustered regularly interspersed short palindromic re-peats [CRISPR]––CRISPR-associated genes) systems of their bacterial hosts for successful invasion and survival. Hence, anti-CRISPRs have great potential to serve as regulators/modulators of CRISPR–Cas genome editing tools for safer and more controllable genome engineering.

While research of anti-CRISPR is very young, it is growing at a remarkable rate. Since their discovery in 2013, there have been >130 papers published. At present, there are few homology-based search tools to detect anti-CRISPRs (e.g. include two online databases **Anti-CRISPRdb** and **CRISPRminer**, as well as a Google Doc for unifying anti-CRISPR nomenclature (https://tinyurl.com/anti-CRISPR)). Clearly, there is a lack of bioinformatics resources for anti- CRISPRs, although most of the 56 experimentally characterized Acr protein families had been identified.

This project intends to use machine learning approaches to identify phage Arc according to their function.

The main goal of this project is:
**1)** To review literature related to annotation tools applied to identify Acr;
**2)** To build a dataset of Arc protein sequences using available datasets and NCBI queries;
**3)** To start exploring machine learning approaches to improve the accuracy of predictions of Arc.
